src/engine/HasRelationScan.cpp

```diff
@@ -0,0 +1,458 @@
+struct resizeIfVec<vector<C>, C> {
+  static void resize(vector<C>& t, int size) { t.resize(size); }
+};
+
+template <typename A, typename R>
+void doComputeSubqueryS(const std::vector<A>* input,
+                        const size_t inputSubjectColumn, std::vector<R>* result,
+                        const std::vector<PatternID>& hasPattern,
+                        const CompactStringVector<Id, Id>& hasRelation,
+                        const CompactStringVector<size_t, Id>& patterns) {
+  for (size_t i = 0; i < input->size(); i++) {
+    const A& inputRow = (*input)[i];
+    size_t id = inputRow[inputSubjectColumn];
+    if (id < hasPattern.size() && hasPattern[id] != NO_PATTERN) {
+      // add the pattern
+      size_t numPredicates;
+      Id* patternData;
+      std::tie(patternData, numPredicates) = patterns[hasPattern[id]];
+      for (size_t i = 0; i < numPredicates; i++) {
+        R row;
+        resizeIfVec<R, Id>::resize(row, inputRow.size() + 1);
+        for (size_t j = 0; j < inputRow.size(); j++) {
+          row[j] = inputRow[j];
+        }
+        row[inputRow.size()] = patternData[i];
+        result->push_back(row);
+      }
+    } else if (id < hasRelation.size()) {
+      // add the relations
+      size_t numPredicates;
+      Id* predicateData;
+      std::tie(predicateData, numPredicates) = hasRelation[id];
+      for (size_t i = 0; i < numPredicates; i++) {
+        R row;
+        resizeIfVec<R, Id>::resize(row, inputRow.size() + 1);
+        for (size_t j = 0; j < inputRow.size(); j++) {
+          row[j] = inputRow[j];
+        }
+        row[inputRow.size()] = predicateData[i];
+        result->push_back(row);
+      }
+    } else {
+      break;
+    }
+  }
+}
+
+/**
+ * @brief  This struct is used to call doComputeSubqueryS with the correct
+ * template parameters. Using it is equivalent to a structure of nested if
+ * clauses, checking if the inputColCount and resultColCount are of a certain
+ * value, and then calling doGroupBy.
+ */
+template <int InputColCount>
+struct CallComputeSubqueryS {
+  static void call(int inputColCount, const ResultTable* input,
+                   const size_t inputSubjectColumn, ResultTable* result,
+                   const std::vector<PatternID>& hasPattern,
+                   const CompactStringVector<Id, Id>& hasRelation,
+                   const CompactStringVector<size_t, Id>& patterns) {
+    if (InputColCount == inputColCount) {
+      if (InputColCount < 5) {
+        result->_fixedSizeData = new vector<array<Id, InputColCount + 1>>();
+        doComputeSubqueryS<std::array<Id, InputColCount>,
+                           std::array<Id, InputColCount + 1>>(
+            static_cast<vector<array<Id, InputColCount>>*>(
+                input->_fixedSizeData),
+            inputSubjectColumn,
+            static_cast<vector<array<Id, InputColCount + 1>>*>(
+                result->_fixedSizeData),
+            hasPattern, hasRelation, patterns);
+      } else {
```

Comment:
Add a comment to make it clear that this is only for the `== 5` case where the input uses `_fixedSizeData` still but the output already uses `_varSizeData`.

Reply:
done